refactor: use the scverse cookiecutter template

ori-kron-wis opened the pull request:
Aligns the repository with the [scverse cookiecutter template](https://github.com/scverse/cookiecutter-scverse) v0.8.0 and tracks it in `.cruft.json`, so future template changes arrive as automated pull requests.

Closes #3600.

## Packaging

- `dev`, `test` and `doc` dependencies move from `[project.optional-dependencies]` to PEP 735 `[dependency-groups]`. The `tests`, `test`, `editing`, `dev`, `docs` and `docsbuild` extras are gone; the feature extras (`autotune`, `cuda`, `hub`, `rapids`, …) stay.
- Every test suite is now a hatch environment in `pyproject.toml`: the `hatch-test` matrix for the default suite, plus `test-cuda`, `test-metal`, `test-dataloaders` and `test-resolution` for the opt-in ones. `pyproject.toml` is the single point of truth — CI runs exactly what `hatch test` runs locally.
- `nvidia-nccl` moved from ad-hoc CI steps into the `cuda`/`cuda13` extras, and `session-info2` was added as a dependency because the new issue form asks reporters for its output.

## CI

- `build.yml` → `build.yaml`, `test_linux.yml` → `test.yaml`, `release.yml` → `release.yaml`, with the template's contents: hatch-driven, `permissions:` blocks, actions pinned by digest, and an `alls-green` check so branch protection needs a single required job.
- The ten specialized workflows were renamed off the `test_linux_*` prefix and rewritten to run through the hatch environments above. Their triggers and label gating are unchanged.
- The release workflow creates the tag with `gh release create` instead of a checkout that persists push credentials, and the Docker matrix uses extras that still exist (`""`, `optional`, `tutorials`).
- Added `.github/dependabot.yml`.

## Tooling

- `.pre-commit-config.yaml` matches the template: biome, `pyproject-fmt` and zizmor replace prettier, mdformat, markdownlint and blacken-docs. `.markdownlint.yaml` is removed; the local "don't commit .rej files" hook stays.
- Added `biome.jsonc` and `.vscode/{settings,launch,extensions}.json`; aligned `.editorconfig`, `.gitignore`, `.readthedocs.yaml` and the issue templates (now GitHub forms).

## Documentation

The docs are now built with `-W`, as the template does. Getting there meant fixing roughly 400 pre-existing warnings rather than muting them:

- Corrected cross-references that pointed at private module paths, wrong modules (`resolvi.train.PyroTrainingPlan`, `pytorch_lightning.trainer.Trainer`) or the wrong role, and fixed the pyro/torch targets to the names those projects actually publish.
- Fixed malformed numpydoc parameter blocks and type names (`np.array` used as a type, `torch.tensor` in annotations, `list[np.array, np.array, np.array]`).
- Documented public classes that were referenced but missing from the API reference, and gave undocumented overrides real docstrings instead of inheriting lightning's.
- Removed dead cross-references from historical `CHANGELOG.md` entries.
- Optional dependencies are now mocked based on whether they actually import, rather than on `READTHEDOCS`, so local, CI and Read the Docs builds behave the same.
- `nitpick_ignore` shrank from 15 entries to 8, all of them third-party or CPython inventory mismatches we cannot address here.

`.readthedocs.yaml` now builds through `uvx hatch run docs:build`. It stays on Python 3.12 rather than the template's 3.14 because the docs pull in the full optional dependency set.

## Notes for reviewers

- Workflow renames change the check names, so branch protection needs updating. `Tests pass in all hatch environments` is the new single required check for the default suite.
- Coverage thresholds and the per-suite flags in `.codecov.yaml` are unchanged; the template's looser defaults were not adopted.
- The tutorials submodule now points at `scverse/scvi-tutorials` rather than the old `YosefLab` URL.

Zethson reviewed (changes requested):
I think that's a start but there's more. I still think that the better way would be to recreate the project from scratch but if this is easier for you, we should ensure that we do it right.

1. What about the issue templates?
2. What about the hatch test set up?
3. What about the whole CI set up? Yours differs greatly from the template and I'm a bit surprised at the number of workflows that you have.
4. The pre-commit setup needs tons of updates, please.
5. The whole configuration of the linters as well -> please also update the pyproject.toml
6. The RTD setup needs updates
7. There's probably more that I missed...

Zethson commented:
@ori-kron-wis where are we with this PR? Should we have an agent finish it? I'd love to see a full version merged.

ori-kron-wis commented:
> @ori-kron-wis, where are we with this PR? Should we have an agent finish it? I'd love to see a full version merged.

Do we have an agent for this specific task? If so, sure, let's run it.

Zethson commented:
Not yet. I can take care of this as it pains me that this isn't done yet. I'll ping you when I'm done.

Zethson committed:
refactor: use the scverse cookiecutter template

Track the scverse cookiecutter template v0.8.0 via .cruft.json and align the
repository with it.

Packaging and CI:

- Move dev, test and documentation dependencies from extras to PEP 735
  dependency groups and declare every test suite as a hatch environment, so
  pyproject.toml is the single point of truth for local and CI runs.
- Rename build.yml, test_linux.yml and release.yml to the template's
  build.yaml, test.yaml and release.yaml, drive all suites through hatch, pin
  actions by digest, add explicit permissions and an alls-green check.
- Create the tag through `gh release create` instead of a credential-persisting
  checkout, and add .github/dependabot.yml.

Tooling:

- Adopt the template's pre-commit config: biome, pyproject-fmt and zizmor
  replace prettier, mdformat, markdownlint and blacken-docs.
- Add biome.jsonc and the .vscode configuration, align .editorconfig,
  .gitignore and .readthedocs.yaml, and convert the issue templates to forms.

Documentation:

- Build with `-W` and fix every warning it surfaced: broken and privately
  pathed cross-references, malformed numpydoc types, undocumented public
  classes and inherited third-party docstrings.

Signed-off-by: Lukas Heumos <lukas.heumos@posteo.net>

Zethson committed:
docs: bump the tutorials submodule

Picks up scverse/scvi-tutorials#562, without which the documentation does not
build with `-W`.

Signed-off-by: Lukas Heumos <lukas.heumos@posteo.net>

Zethson committed:
fix: import the real autotune classes on Read the Docs

scvi.autotune replaced AutotuneExperiment, ScibTuneReportCheckpointCallback and
run_autotune with placeholders whenever READTHEDOCS was set, even though the
autotune dependencies are installed there. The documented classes were the
placeholders, so references to the real class and to its `result_grid`
attribute could not resolve.

Drop the branch and rely on the ImportError fallback that already handles
genuinely missing dependencies, and remove the duplicate placeholder that
docs/conf.py installed.

Signed-off-by: Lukas Heumos <lukas.heumos@posteo.net>

Zethson commented:
@ori-kron-wis there might still be minor docs issues but I'll fix these. Generally, you can already have a look now, please. I know it's a lot of changes but I'm confident that this will help contributions and generally the software quality of scvi-tools (not saying it bad was but the new automations are nice)

Zethson committed:
test: check that scvi.autotune re-exports the real implementations

Guards the regression the previous commit fixed: the placeholders must only be
used when the optional dependencies are missing.

The test lives outside tests/autotune because everything under that directory
carries the `autotune` keyword and is skipped without --autotune-tests.

Signed-off-by: Lukas Heumos <lukas.heumos@posteo.net>

ori-kron-wis commented:
@Zethson, note I updated the tutorials sub-module on main (after recent mergers on scvi-tutorials). Please update it here (currently in conflict and we need to see the updated docs renders)

Zethson committed:
docs: bump the tutorials submodule to match main

Point docs/tutorials/notebooks at e0775d8, the same commit main already
records, so merging this branch does not conflict on the submodule.

Also drop tests/test_autotune_exports.py.

Signed-off-by: Lukas Heumos <lukas.heumos@posteo.net>

Zethson committed:
Merge branch 'main' into Ori-3600-updated-scverse-template

Zethson commented:
@ori-kron-wis updated the submodule. Please have a look. Thank you!

ori-kron-wis commented:
Because of so many changes, well have to test everything

ori-kron-wis commented on docs/conf.py:
```diff
@@ -1,15 +1,18 @@
 from datetime import datetime
 from typing import TYPE_CHECKING
 
+from sphinx.deprecation import RemovedInSphinx90Warning
```
1. (High — latent build-breaker) docs/conf.py imports sphinx.deprecation.RemovedInSphinx90Warning unguarded to suppress a hoverxref warning. That symbol only exists pre-Sphinx-9; once the       
  sphinx>=8.1,<9 pin is ever bumped (which this same PR's cruft-tracking sets up to happen automatically), conf.py fails to import and the whole doc build hard-crashes. No try/except, unlike the  
  new _importable() helper a few lines below for other optional deps.

ori-kron-wis committed:
small updates

ori-kron-wis commented:
close https://github.com/scverse/scvi-tools/pull/3937

ori-kron-wis committed:
update ref.bib

Zethson commented:
Looks like everything is green 🚀

ori-kron-wis committed:
Merge branch 'main' into Ori-3600-updated-scverse-template
